File: GrebFlask/pools.py
from bson import ObjectId
from bson.json_util import loads
from flask import Blueprint, session, render_template

from database import DB, FindCurrentWeek, FindPoolInfoByPooler, FindPoolPicksForWeek, FindPoolPicksForSeason
from football import GetWeek, GetTeamShortName, GetWeekLongName

PoolsBlueprint = Blueprint('pools_blueprint', __name__)

#TODO: Remove test user
poolerid = ObjectId('5f70f0ffd8e2db255c9a0df6')

@PoolsBlueprint.route('/pools')
def index():
    [season, week] = FindCurrentWeek()
    return CreateWeekData(season, week)

@PoolsBlueprint.route('/pools/<strseason>/<strweek>')
def get(strseason, strweek):
    season = int(strseason)
    week = int(strweek)
    return CreateWeekData(season, week)

@PoolsBlueprint.route('/pools/<strseason>')
def getSeason(strseason):
    season = int(strseason)
    return CreateSeasonData(season)

def CreateWeekData(season, week):
    matchdata = GetWeek(season, week)
    matchids = [m['idEvent'] for m in matchdata]

    # pooler = loads(session['pooler'])
    pooler = DB.poolers.find({ '_id': poolerid })[0]
    (poolinfo, poolers) = FindPoolInfoByPooler(pooler)

    weekresults = CalcPoolResults(
        matchdata,
        FindPoolPicksForWeek(season, week, poolers, matchids),
        week
    )

    return {
        'pooldata': poolinfo,
        'poolernames': { str(p['_id']):p['name'] for p in poolers },
        'weekdata': { 'season': season, 'week': week },
        'matches': matchdata,
        'results': weekresults,
    }

def CreateSeasonData(season):
    # pooler = loads(session['pooler'])
    pooler = DB.poolers.find({ '_id': poolerid })[0]
    (poolinfo, poolers) = FindPoolInfoByPooler(pooler)

    seasonPicks = FindPoolPicksForSeason(season, poolers)
    results = [ CalcPoolResults(GetWeek(season, w+1), picks, w+1) for w, picks in enumerate(seasonPicks) ]
    totals = [ { res['pid'] : res['total'] for res in result } for w, result in enumerate(results) ]
    fulltotals = { str(pooler['_id']) : 0 for pooler in poolers }

    for t in totals:
        for pid, score in t.items():
            fulltotals[pid] = fulltotals[pid] + score

    return {
        'pooldata': poolinfo,
        'poolernames': { str(p['_id']):p['name'] for p in poolers },
        'seasoninfo': season,
        'totals': totals,
        'fulltotals': fulltotals,
    }

def CalcPoolResults(matches, poolpicks, week):
    results = []

    for poolerid, picks in poolpicks.items():
        total = 0
        res = {}

        for match in matches:
            hscore = int(match['intHomeScore']) if match['intHomeScore'] is not None else 0
            ascore = int(match['intAwayScore']) if match['intAwayScore'] is not None else 0

            hwin = hscore > ascore
            tied = hscore == ascore

            hpick = picks[match['idEvent']] == GetTeamShortName(match['strHomeTeam'])
            rightpick = (hwin and hpick) or (not(hwin) and not(hpick))

            others = [ p[match['idEvent']] if i != poolerid else '' for i, p in poolpicks.items() ]
            unique = not(picks[match['idEvent']] in others)

            if tied:
                score = 1
            elif rightpick and unique:
                score = int(GetCorrectScore(week) * 1.5)
            elif rightpick:
                score = GetCorrectScore(week)
            else:
                score = 0

            res[match['idEvent']] = {
                'pick': picks[match['idEvent']],
                'score': score,
            }

            total = total + score

        results.append({ 'pid': poolerid, 'scores': res, 'total': total })

    return results

def GetCorrectScore(week_num):
    if (type(week_num) == str):
        n = int(week_num)
    else:
        n = week_num

    if   n == 19: return 4
    elif n == 20: return 6
    elif n == 21: return 8
    elif n == 22: return 10
    else:         return 2

File: GrebReact/src/utils/pools.js
import { FindPoolInfoByPooler, FindPoolPicksForWeek, FindPoolPicksForSeason } from '../lib/database'
import { GetWeek, GetTeamShortName } from './football'

export async function CreateWeekData(season, week, pooler) {
    const matchdata = await GetWeek(season, week)
    const matchids = matchdata.map((m) => m['idEvent']);

    const [poolinfo, poolers] = await FindPoolInfoByPooler(pooler)
    let names = {};
    poolers.forEach((p) => names[String(p['_id'])] = p['name']);

    const picks = await FindPoolPicksForWeek(season, week, poolers, matchids);
    const weekresults = CalcPoolResults( matchdata, picks, week);

    return {
        'pooldata': poolinfo,
        'poolernames': names,
        'weekdata': {'season': season, 'week': week},
        'matches': matchdata,
        'results': weekresults,
    };
}

export async function CreateSeasonData(season, pooler) {
    const [poolinfo, poolers] = await FindPoolInfoByPooler(pooler)
    let names = {};
    poolers.forEach((p) => names[String(p['_id'])] = p['name']);

    const seasonpicks = await FindPoolPicksForSeason(season, poolers)
    let results = [];
    for (let w in seasonpicks) {
        const week = Number(w);
        const matchdata = await GetWeek(season, week+1);
        results.push(CalcPoolResults(matchdata, seasonpicks[w], week+1));
    }

    let totals = [];
    for (let w in results) {
        const result = results[w];
        let temp = {};
        result.forEach((r) => temp[r['pid']] = r['total']);
        totals.push(temp);
    }
    let fulltotals = {};
    poolers.forEach((p) => fulltotals[String(p['_id'])] = 0);

    for (let i in totals) {
        for (let p in totals[i]) {
            fulltotals[p] += totals[i][p];
        }
    }

    return {
        'pooldata': poolinfo,
        'poolernames': names,
        'seasoninfo': season,
        'totals': totals,
        'fulltotals': fulltotals,
    };
}

function CalcPoolResults(matches, poolpicks, week) {
    let results = []

    for (let key in poolpicks) {
        let total = 0;
        let res = {};

        for (let i in matches) {
            const match = matches[i];

            let hscore = 0;
            if (match['intHomeScore']) { hscore = Number(match['intHomeScore']); }
            let ascore = 0;
            if (match['intAwayScore']) { ascore = Number(match['intAwayScore']); }

            const hwin = hscore > ascore;
            const tied = hscore == ascore;

            const hpick = poolpicks[key][match['idEvent']] == GetTeamShortName(match['strHomeTeam']);
            const rightpick = (hwin && hpick) || (!hwin && !hpick);

            let others = [];
            for (let o in poolpicks) {
                if (o != key) {
                    others.push(poolpicks[o][match['idEvent']]);
                }
            }
            const unique = !others.includes(poolpicks[key][match['idEvent']])

            let score = 0;
            if (tied) {
                score = 1;
            }
            else if (rightpick && unique) {
                score = Number(GetCorrectScore(week) * 1.5);
            }
            else if (rightpick) {
                score = GetCorrectScore(week);
            }
            else {
                score = 0;
            }

            res[match['idEvent']] = {
                'pick': poolpicks[key][match['idEvent']],
                'score': score,
            };
            total += score
        }

        results.push({ 'pid': key, 'scores': res, 'total': total })
    }

    return results
}

function GetCorrectScore(week_num) {
    let n = 0;
    if (typeof week_num == String) {
        n = int(week_num)
    }
    else {
        n = week_num
    }

    if      (n == 19) { return 4; }
    else if (n == 20) { return 6; }
    else if (n == 21) { return 8; }
    else if (n == 22) { return 10; }
    else              { return 2; }
}
